lib/IDE/Formatting.cpp

```diff
@@ -115,761 +277,2279 @@ class FormatContext {
+  bool overlapsTarget(SourceLoc Start, SourceLoc End) const {
+    assert(Start.isValid());
+    return !SM.isBeforeInBuffer(TargetLocation, Start) &&
+      (End.isInvalid() || !SM.isBeforeInBuffer(End, TargetLocation));
+  }
+
+  /// Checks whether the given range overlaps the target location.
+  ///
+  /// \return \c Range.Start <= \c TargetLocation <= \c Range.End
+  bool overlapsTarget(SourceRange Range) const {
+    assert(Range.isValid());
+    return overlapsTarget(Range.Start, Range.End);
+  }
+
+  /// Checks whether the given range contains the target location.
+  ///
+  /// \return \c Start < \c TargetLocation < \c End
+  bool containsTarget(SourceLoc Start, SourceLoc End) const {
+    assert(Start.isValid());
+    return SM.isBeforeInBuffer(Start, TargetLocation) &&
+      (End.isInvalid() || SM.isBeforeInBuffer(TargetLocation, End));
+  }
+
+  /// Checks whether the given range contains the target location.
+  ///
+  /// \return \c Range.Start < \c TargetLocation < \c Range.End
+  bool containsTarget(SourceRange Range) const {
+    assert(Range.isValid());
+    return containsTarget(Range.Start, Range.End);
+  }
+
+#pragma mark Declaration indent contexts
+
+  Optional<IndentContext>
+  getIndentContextFrom(Decl *D, Optional<Trailing> TrailingTarget) {
+
+    if (auto *AFD = dyn_cast<AbstractFunctionDecl>(D)) {
+      SourceLoc ContextLoc = AFD->getStartLoc();
+      // If this is a getter without a 'get' loc, the context loc is the start
+      // of its storage.
+      if (auto *AD = dyn_cast<AccessorDecl>(AFD)) {
+        if (AD->isGetter() && AD->getAccessorKeywordLoc().isInvalid()) {
+          auto *ASD = AD->getStorage();
+          if (auto *VD = dyn_cast_or_null<VarDecl>(ASD)) {
+            ContextLoc = VD->getStartLoc();
+          } else if (auto *SD = dyn_cast_or_null<SubscriptDecl>(ASD)) {
+            ContextLoc = SD->getStartLoc();
+          }
+        }
+      }
+      if (auto Ctx = getIndentContextFrom(AFD->getBody(), ContextLoc))
+        return Ctx;
+      if (auto Ctx = getIndentContextFrom(AFD->getParameters(), ContextLoc))
+        return Ctx;
+      if (auto Ctx = getIndentContextFrom(AFD->getGenericParams(), ContextLoc, D))
+        return Ctx;
+
+      if (TrailingTarget)
+        return None;
+      return IndentContext {ContextLoc, false};
+    }
+
+    if (auto *NTD = dyn_cast<NominalTypeDecl>(D)) {
+      SourceLoc ContextLoc = NTD->getStartLoc();
+
+      if (auto Ctx = getIndentContextFromInherits(NTD->getInherited(), ContextLoc))
+        return Ctx;
+      if (auto Ctx = getIndentContextFromBraces(NTD->getBraces(), ContextLoc, NTD))
+        return Ctx;
+      if (auto Ctx = getIndentContextFrom(NTD->getGenericParams(), ContextLoc, D))
+        return Ctx;
+      if (auto Ctx = getIndentContextFrom(NTD->getTrailingWhereClause(), ContextLoc, D))
+        return Ctx;
+
+      if (TrailingTarget)
+        return None;
+      return IndentContext {ContextLoc, false};
```

Comment:
When would we fall through to this?

Reply:
On anything in-between the ranges handled by the above, e.g. the comment below:
```
struct Foo<T, U, V>
// FIXME: some important comment about the constraints below
where
    T: Equatable,
    U: Equatable,
    V: Codable
{
}
```

Reply:
Ahhhh, right thanks! I forgot the ranges wouldn't necessarily be continuous.